proc/proc.go

```diff
@@ -0,0 +1,249 @@
+// proc offers control over arbitrary processes.
+// proc depends heavily on os/exec.
+// If/when proc needs to do resource control & isolation
+// its probable that os/exec will need to be replaced with
+// either the os package itself or syscall.
+package proc
+
+// Fingers crossed that its easier than rewriting os.ForkExec
+
+import (
+	"bufio"
+	"context"
+	"fmt"
+	"io/ioutil"
+	"os"
+	"os/exec"
+	"sync"
+	"time"
+)
+
+const outputReadRefreshInterval = time.Millisecond * 10
+
+// Proc is a BitBox process.
+// Stderr and Stdout are dumped to temp files on disk.
+type Proc struct {
+	// TODO: there is 99% chance we need a synchronization primitive at this level
+	stdoutFileName string
+	stderrFileName string
+	cmd            *exec.Cmd
+}
+
+// NewProc constructs and begins process.
+// Stdout & Stderr of the new process are pointed at temp files.
+// The tempfileNames are acessable through the coresponding members.
+func NewProc(cmdName string, args ...string) (Proc, error) {
+
+	var err error
+	var cmdPath string
+	if cmdPath, err = exec.LookPath(cmdName); err != nil {
+		return Proc{}, err
+	}
+	cmd := exec.Command(cmdPath, args...)
+
+	var stdout *os.File
+	if stdout, err = ioutil.TempFile("", ""); err != nil {
+		return Proc{}, err
+	}
+	cmd.Stdout = stdout
+	stdoutName := stdout.Name()
+
+	var stderr *os.File
+	if stderr, err = ioutil.TempFile("", ""); err != nil {
+		return Proc{}, err
+	}
+	cmd.Stderr = stderr
+	stderrName := stderr.Name()
+
+	if err = cmd.Start(); err != nil {
+		return Proc{}, err
+	}
+
+	go func() {
+		// Wait on the cmd to make sure resources get released
+		cmd.Wait()
+	}()
+
+	return Proc{
+		stdoutFileName: stdoutName,
+		stderrFileName: stderrName,
+		cmd:            cmd,
+	}, nil
+}
+
+// Kill causes the running process to exit and closes the related resources.
+func (p Proc) Kill() error {
+	if err := p.cmd.Process.Kill(); err != nil {
```

Comment:
it would be nice if we'd only attempt to kill running processes

Reply:
Indeed, idempotency (at least at the API level) would be nice.

Reply:
> Is there an advantage that you feel is important?

Presently, `err` could mean that something else called `Kill` first (mostly harmless), or it could mean that `Kill` failed and the job is still running (requires followup action from the user). There isn't presently a way to know without some reconciliation logic (error parsing or calling Query). This logic needs to be written somewhere.  Are you going to make anyone using your cli tool handle those two cases, or write a shim at some layer?  If so, which layer makes more sense than this one?

Reply:
TBF, I don't love the racy "check if it is alive before killing", better to try and have smart error disambiguation IMO.

Reply:
>it could mean that Kill failed and the job is still running

I think I am missing how this could happen. I want to make it clear that I don't know everything in the world about syscalls. With that said, the documentation for [kill(2)](https://man7.org/linux/man-pages/man2/kill.2.html), which is called by `Process.Kill()`, documents only 3 errors, and none of them imply to me that the process could have survived. 

>EINVAL An invalid signal was specified.

>EPERM  The calling process does not have permission to send the signal to any of the target processes.

>ESRCH  The target process or process group does not exist.  Note that an existing process might be a zombie, a process that has terminated execution, but has not yet been wait(2)ed for.

Changing the userID of the process could escape that, but I don't think I am trying to worry about that quite yet.

Reply:
If the thing you are looking for is typing of the errors instead of relying on the string errors returned by `Process.Kill()` I can understand that.

Maybe you are suggesting that I keep track of whether or not this function has been called once? I can understand why you might find that useful for the disambiguation of errors for the user. I shy away from that because the functionality seems so similar to the error handling/response provided by the underlying libraries.

Are you just suggesting that we don't pass through errors that communicated something like "this process is already stopped"? If that's the case, this conversation and the one below has me leaning to just renaming the function to `Stop()` to change the expectations around what an error from this function means.

Reply:
I opted to change the `Kill()` to `Stop()` because of the bellow comment. If that resolves this conversation great, otherwise I'd love a little more direction from you @wadells and @NajiObeid 😄

Reply:
> Maybe you are suggesting that I keep track of whether or not this function has been called once?
> Are you just suggesting that we don't pass through errors that communicated something like "this process is already stopped"?
>  renaming the function to Stop() to change the expectations around what an error from this function means.

Do what you think is best.  As before, I encourage impotency for this call at the grpc API level because I believe it is a better UX.

Reply:
Walt you didn't really mean impotency in your last comment 😄

Reply:
https://play.golang.org/p/qCoZFh_OGPN

This doesn't look particularly right to me personally

Reply:
I'm glad you saw that @NajiObeid. @wadells comment cracked me up.

Wow! Thank you so much for that playground example. This really helped me see what your concern is. A snippet says a thousand words, or ya know, something like that.

I've made two more examples to more closely represent the functionality in this package. A new go routine with a call to `cmd.Wait()` is started/scheduled before `NewProc()` returns the new process and by extension the processID to the client. 

1. [Kill before the process would normally exit](https://play.golang.org/p/9VM6VhYCzDE) 
2. [Kill after the process has exited](https://play.golang.org/p/zns6vUQIYbb) 

This isn't perfectly idempotent because the calls to `Kill()` do display a sense of state in terms of a running vs non-running process. With that said, it seems like the intended behavior to have the API tell you that the process is already finished.

I could definitely see something showing up in the future that requires a little more intelligence here. For now, I think between the name change and seeing the expected operation I'm ok to move on. Do you each feel about the same?